Documentation/install/upgrade.rst

```diff
@@ -257,9 +257,7 @@ IMPORTANT: Changes required before upgrading to 1.7.0
 
   .. code:: bash
 
-     $ kubectl get ds -n kube-system cilium -o yaml > cilium-ds.yaml
-     $ sed -i '/kubernetes.io\/cluster-service: "true"/d' cilium-ds.yaml
-     $ kubectl apply -f cilium-ds.yaml --force
+     $ kubectl -n kube-system patch daemonset cilium --type=json -p='[{"op": "remove", "path": "/spec/selector/matchLabels/kubernetes.io~1cluster-service"}]'
```

Comment:
When I tried this, I hit this problem:

```
$ kubectl -n kube-system patch daemonset cilium --type=json -p='[{"op": "remove", "path": "/spec/selector/matchLabels/kubernetes.io~1cluster-service"}]'
The DaemonSet "cilium" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"k8s-app":"cilium"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
```

As far as I can tell, there is no corresponding `--force` option.

Reply:
That's not what we want.

I run kubectl Client-version v1.14.1 and I tried on k8s:
- GCP (GCE): v1.14.8-gke.17
- AWS (EKS): v1.14.7-eks-1861c5
Upgrading from Cilium v1.5.9 to 1.7.0-RC1 and both envs works correctly.
As far as I know you don't have the --force bool in the patch command.

Which environment are you testing?

Reply:
Removed `release-note` section from description, sorry about this.

Reply:
I was trying in a microk8s environment, k8s v1.16.

This "field is immutable" message is the same one that I get if I manually do `kubectl -n kube-system edit ds cilium` and delete the lines referencing `cluster-service`. It's the same base reason why this upgrade step is necessary in the first place; if this section was mutable then we can just apply the daemonset on top without the `--force`.

The other thing that I'm not sure about, is that there's three references to `cluster-service` in the DS YAML, does this patch command remove all three?
* .metadata.labels
* .spec.selector.matchLabels
* .spec.template.metadata.labels

Reply:
No it removed only the 2nd, which is fine imo, because otherwise - because of 3rd bullet - the DS goes to restart the existing Pods with the new labels (which is unwanted for me, because I want to update them to a newer version). Only the 2nd results in the 'field is immutable' error the others can be updated by the new yaml without any issues (only normal labels).

Note 1) If I do not patch the cilium DS, I got the error during upgrade that the 'field is immutable', so a prepare step is needed.
Note 2) Doing an edit on the cilium ds and remove the field results in the same 'field is immutable' error during save.

Maybe the behavior is k8s 1.14 and 1.16 is slightly different and my patch doesn't work in 1.16 anymore?

Reply:
Interesting, thanks for the explanation.

I think the 1.14/1.16 version difference theory is the best we can go on right now. I can ask around and see if I can get access to other versions to try out; I'm also quite happy to try out any alternative commands in my environment. Feel free to reach out to me on Slack if it helps to close the feedback loop here.

Reply:
I think this may be related to the DaemonSet storage version in 1.14. Can you confirm you're on `apps/v1` @Robert-Stam?
Kubernetes 1.16 defaults to `apps/v1` which now enforces the immutability of this field.

Reply:
An alternative could be ```shell
$ kubectl get ds -n kube-system cilium -o yaml > cilium-ds.yaml
$ kubectl delete ds -n kube-system cilium --cascade=false
$ sed -i '/kubernetes.io\/cluster-service: "true"/d' cilium-ds.yaml
$ kubectl create -f cilium-ds.yaml
```

This doesn't simplify it but works.

Reply:
We are using Cilium v1.5.9 on k8s v1.14, which has:
```apiVersion: extensions/v1beta1```
So this explains

Reply:
I checked AKS (Azure) as well, there we use k8s v1.15.7, however there it's still
```apiVersion: extensions/v1beta1```

So my suggestion will not work correctly as of k8s v1.16, so I suggest to close this ticket.

Reply:
I still think there's value in fixing the instructions as they are, because I think they won't work for the same reasons your simplification didn't.

Reply:
@pires are you saying that you tried the current instructions and they do not work for you?

I have used the current instructions (note the `--force`) in a few upgrade scenarios under the microk8s 1.16 environment I use for development and it works for me.

I suspect that the DS delete `--cascade=false` is just a slightly different alternative to the `kubectly apply ... --force`, so it seems neither here nor there for which approach to take. But I'd like to confirm if you're observing other behaviour.

Reply:
No, I didn't so maybe it works.

Reply:
OK, in that case if you do end up doing any testing on this front and confirm that it either works or doesn't work, then please do reach out or file an issue so we can make sure it will work. Otherwise we'll just rely on the standard release testing / community feedback for validating this.

Good spotting on the `apiVersion` by the way, thanks for the feedback there!